src/Microsoft.AspNet.Mvc.Core/ControllerActionDescriptorBuilder.cs

```diff
@@ -284,18 +285,37 @@ private static ParameterDescriptor CreateParameterDescriptor(ParameterModel para
 
         private static void AddApiExplorerInfo(
             ControllerActionDescriptor actionDescriptor,
-            ActionModel action,
-            ControllerModel controller)
+            ApplicationModel application,
+            ControllerModel controller,
+            ActionModel action)
         {
-            var apiExplorerIsVisible = action.ApiExplorer?.IsVisible ?? controller.ApiExplorer?.IsVisible ?? false;
-            if (apiExplorerIsVisible)
+            
+            var isVisible = 
+                action.ApiExplorer?.IsVisible ?? 
+                controller.ApiExplorer?.IsVisible ?? 
+                application.ApiExplorer.IsVisible ??
+                false;
+
+            var isVisibleSetOnActionOrController = 
+                action.ApiExplorer?.IsVisible ?? 
+                controller.ApiExplorer?.IsVisible ?? 
+                false;
+
+            var isVisibleSetOnApplication = application.ApiExplorer?.IsVisible ?? false;
+
+            if (isVisibleSetOnActionOrController && !IsAttributeRoutedAction(actionDescriptor))
             {
-                if (!IsAttributeRoutedAction(actionDescriptor))
-                {
-                    // ApiExplorer is only supported on attribute routed actions.
-                    throw new InvalidOperationException(Resources.FormatApiExplorer_UnsupportedAction(
-                        actionDescriptor.DisplayName));
-                }
+                // ApiExplorer is only supported on attribute routed actions.
+                throw new InvalidOperationException(Resources.FormatApiExplorer_UnsupportedAction(
+                    actionDescriptor.DisplayName));
+            }
+            else if (isVisibleSetOnApplication && !IsAttributeRoutedAction(actionDescriptor))
```

Comment:
It's probably better to explain the just ignore it, something like - Since we don't support ApiExplorer for convetional routes,we will just skip this action and it will not show up in ApiExplorer.